pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -178,92 +190,257 @@ public CompletableFuture<Void> deleteTopicPoliciesAsync(TopicName topicName,
+                                                return CompletableFuture.completedFuture(null);
+                                            }
+                                            TopicPolicies policiesToUpdate;
+                                            if (actionType == ActionType.DELETE) {
+                                                policiesToUpdate = null; // For delete, policies object is null
                                             } else {
-                                                result.completeExceptionally(
-                                                        new RuntimeException("Got message id is null."));
+                                                policiesToUpdate = currentPolicies.isEmpty()
+                                                        ? createTopicPolicies(isGlobalPolicy)
+                                                        : currentPolicies.get().clone();
+                                                policyUpdater.accept(policiesToUpdate);
                                             }
-                                        }
-                                    });
-                            }
-                    });
-                    return result;
-                });
+                                            return sendTopicPolicyEventInternal(topicName, actionType, policiesToUpdate,
+                                                    isGlobalPolicy);
+                                        })
+                                        .thenCompose(messageId -> {
+                                            if (messageId == null) {
+                                                return CompletableFuture.completedFuture(null);
+                                            } else {
+                                                // asynchronously wait until the message ID is read by the reader
+                                                return untilMessageIdHasBeenRead(topicName.getNamespaceObject(),
+                                                        messageId);
+                                            }
+                                        });
+                            }));
+        }).whenComplete((res, ex) -> {
+            // remove the current future from the sequencer map, if it is done
+            // this would remove the future from the sequencer map when the last operation completes in the chained
+            // future
+            topicPolicyUpdateSequencer.compute(sequencerKey, (key, existingFuture) -> {
+                if (existingFuture != null && existingFuture.isDone()) {
+                    // Remove the completed future from the sequencer map
+                    return null;
+                }
+                return existingFuture;
+            });
+            if (ex != null) {
+                writerCaches.synchronous().invalidate(topicName.getNamespaceObject());
```

Comment:
Remove the writer cache isn't right here, I think it should be https://github.com/apache/pulsar/pull/24427/files#r2297933245

Reply:
oh yes, checking again

Reply:
resolved